A gen.Registrar implementation for Apache ZooKeeper

qjpcpu opened the pull request:
Work just like etcd registrar.

qjpcpu committed:
feat: zk registrar

halturin commented:
I'm glad you're interested in the framework and want to make it better. But, no offense - what you’ve produced feels more like some vibed-out draft than something functional.
I don’t see how node registration is made atomic, and I don’t see any cluster-level events regarding nodes.
If you genuinely want to contribute, please take some time to study the implementations of registrars like Saturn and etcd, considering all aspects of their functionality.
I would truly appreciate it if you could bring this project to a fully robust, production-grade state.

PS: dont see any reason to use your personal  fork of go-zookeeper as a dependency

qjpcpu commented:
> I'm glad you're interested in the framework and want to make it better. But, no offense - what you’ve produced feels more like some vibed-out draft than something functional. I don’t see how node registration is made atomic, and I don’t see any cluster-level events regarding nodes. If you genuinely want to contribute, please take some time to study the implementations of registrars like Saturn and etcd, considering all aspects of their functionality. I would truly appreciate it if you could bring this project to a fully robust, production-grade state.
> 
> PS: dont see any reason to use your personal fork of go-zookeeper as a dependency

@halturin 

Sorry for my brief comments. This pr is truely in a production state, those code are validated in a big project of my company.
* [how node register] when node start up, the registrar will create a ProtectedEphemeralSequential znode on zk, when node disappear, the znode will be deleted automatically. 
* [how node discover] when node start up, the registrar will discover all nodes in the cluster. then, a watcher would be created to watch the cluster znode. A EventNodeJoined event would be sent when new node join the cluster.A NodeLeave event would be sent when a node leave the cluster.

Besides, I introduce some new feature that etcd register not support.
* [NAT env support] user can start registrar with a advertised address with WithAdvertiseAddress(), then registrar will use the advertised ip to register to zk. so cluster peers would connect each other smothly.
* [custom zk dialer] in a private infra environment such as my company, user can't dial zk directly, so registrar need a custom zk dialer to dial zk.
* [toggle for discover self] I think registrar should return all nodes/app_routes, keep the right for user to decide how to use it, but etcd registrar always drop self when resolve. So I build a toggle `ExcludeSelfWhenResolve`.

And, why I use my personal fork go-zookeeper, cause I wanna expose raw zk connection options to user, but original go-zookeeper define `connOption` as private, oops.

Later, maybe I can make a pr to https://github.com/ergo-services/examples, add more examples about how to use this zk registrar.
> A lot of the implementation for this feature was heavily borrowed from your work in etcd registrar. thanks a lot!

halturin commented:
@qjpcpu  thank you for the clarification. I'll take a deeper look. Anyway, the current implementation looks messy, and I would kindly ask you to clean it up.

According to your extra features:
 - [toggle for discover self] - It was made intentionally. There is no need to ask the registrar if you have a local instance of the application. It isn't very sensible.
 - [NAT env support] - it's in my roadmap (to extend AcceptorOptions), which will allow adding "NATed" acceptors. But it has to be designed with not to break anything.

halturin commented:
Regarding code style: look at the other repos, pay attention to naming, especially - how usually Options are declared (you may notice, we use straightforward option setting - via Options struct).

qjpcpu committed:
1. code style: use Options struct instead of function options
2. remove ExcludeSelfWhenResolve toggle, align with author's design
3. add more ut

qjpcpu commented:
@halturin I refact my pr, follow your suggestions. appreciate that if you review the new commit.
1. use struct Options
2. follow your design to remove ExcludeSelfWhenResolve
3. add more ut & rewrite readme

halturin commented on zk/README.md:
```diff
@@ -0,0 +1,164 @@
+
+var leaderPID gen.PID
+
+roleListener := func(role zk.RoleType) {
```
use registrar's event for that, so any process subscribed to it can change its behavior.

halturin commented on zk/options.go:
```diff
@@ -0,0 +1,98 @@
+	// environments (e.g., Docker, Kubernetes).
+	RoutesMapper RoutesMapper
+
+	// LogFn is a custom logging function for internal registrar output. Defaults to a silent logger.
```
must be similar to etcd/Saturn - logging has t obe done via Node's logging subsystem https://devel.docs.ergo.services/basics/logging

halturin commented on zk/options.go:
```diff
@@ -0,0 +1,98 @@
+	LogFn LogFn
+
+	// ZkOptions allows passing native go-zk library connection options for advanced configuration.
+	ZkOptions []zk.ConnOption
```
not sure about this. as I see, you need it to customize Dialer, right? There should be another way.

halturin commented:
I did a partial review, and at some point I realized that you’re solving too many very specific cases in this project (patched go-zookeeper, using leader election, NATed nodes).
I understand why you brought these things in, but I don’t see a reason to have them in the ZooKeeper registrar implementation for everyone — it’s far too case-specific.
Aside from that, I find the leader-election functionality unsuitable for a registrar.
(With the 3.2 release, actor.Raft will also be available, so anyone will be able to create actor groups with leader election anyway.)
NAT support also shouldn’t be a hack — and in this context it basically is.
Given this set of constraints, it seems unlikely that you’ll end up with a version that could be included as part of the official extra library.
Maybe it would make more sense to publish it as a separate, third-party library in your own profile?

qjpcpu commented:
> I did a partial review, and at some point I realized that you’re solving too many very specific cases in this project (patched go-zookeeper, using leader election, NATed nodes). I understand why you brought these things in, but I don’t see a reason to have them in the ZooKeeper registrar implementation for everyone — it’s far too case-specific. Aside from that, I find the leader-election functionality unsuitable for a registrar. (With the 3.2 release, actor.Raft will also be available, so anyone will be able to create actor groups with leader election anyway.) NAT support also shouldn’t be a hack — and in this context it basically is. Given this set of constraints, it seems unlikely that you’ll end up with a version that could be included as part of the official extra library. Maybe it would make more sense to publish it as a separate, third-party library in your own profile?

@halturin 

That sounds very reasonable. I'll close this PR then. Thank you for your time.

qjpcpu commented:
@halturin BTW, you create a really great project. ergo is very clean and powerful 👍
